pkg/contract-manager/reservation_controller.go

```diff
@@ -47,13 +55,309 @@ type ReservationReconciler struct {
+
+			case nodecorev1alpha1.PhaseSolved:
+				klog.Infof("Reserve phase for the reservation %s solved", reservation.Name)
+				return ctrl.Result{}, nil
+			}
+		}
+
+		if reservation.Spec.Purchase {
+			switch reservation.Status.PurchasePhase.Phase {
+			case nodecorev1alpha1.PhaseIdle:
+				// Purchase the flavour with the given TransactionID
+				resPur, err := purchaseFlavour(ctx, transaction, &reservation)
+				if err != nil {
+					klog.Errorf("Error when purchasing flavour for Reservation %s: %s", req.NamespacedName, err)
+					setPurchaseStatus(&reservation, nodecorev1alpha1.PhaseFailed)
+					return ctrl.Result{}, err
+				}
+				klog.Infof("Reservation %s for flavour for buyer %s", resPur.FlavourID, resPur.BuyerID)
+				klog.Infof("Purchase completed with status %s", resPur.Status)
+
+			case nodecorev1alpha1.PhaseRunning:
+				// Update the spec of the reservation
+				reservation.Spec.TransactionID = transaction.TransactionID
+				if err := r.Update(ctx, &reservation); err != nil {
+					klog.Errorf("Error when updating Reservation %s: %s", reservation.Name, err)
+					setPurchaseStatus(&reservation, nodecorev1alpha1.PhaseFailed)
+					return ctrl.Result{}, err
+				}
+				klog.Infof("Reservation %s updated", reservation.Name)
+
+				setReservationPhase(&reservation, nodecorev1alpha1.PhaseSolved)
+
+				if err := r.updateReservationStatus(ctx, &reservation); err != nil {
+					klog.Errorf("Error when updating Reservation %s status: %s", req.NamespacedName, err)
+					setPurchaseStatus(&reservation, nodecorev1alpha1.PhaseFailed)
+					return ctrl.Result{}, err
+				}
+
+				// Create a contract CR now that the reservation is solved
+				contract := createContractCustomResource(ctx, &reservation)
+				if err := r.Create(ctx, contract); err != nil {
+					klog.Errorf("Error when creating Contract %s: %s", contract.Name, err)
+					setPurchaseStatus(&reservation, nodecorev1alpha1.PhaseFailed)
+					return ctrl.Result{}, err
+				}
+				klog.Infof("Contract %s created", contract.Name)
+
+				setPurchaseStatus(&reservation, nodecorev1alpha1.PhaseSolved)
+
+				// Update the status for reconciliation
+				if err := r.updateReservationStatus(ctx, &reservation); err != nil {
+					klog.Errorf("Error when updating Reservation %s status: %s", req.NamespacedName, err)
+					return ctrl.Result{}, err
+				}
+
+			case nodecorev1alpha1.PhaseFailed:
+				klog.Infof("Purchase phase for the reservation %s failed", reservation.Name)
+				return ctrl.Result{}, nil
+
+			case nodecorev1alpha1.PhaseSolved:
+				klog.Infof("Purchase phase for the reservation %s solved", reservation.Name)
+				return ctrl.Result{}, nil
+			}
+		}
+		return ctrl.Result{}, nil
+	case nodecorev1alpha1.PhaseSolved:
+		klog.Infof("Reservation %s solved", reservation.Name)
+		// TODO: for Ale, create the Allocation Object
+		return ctrl.Result{}, nil
+	case nodecorev1alpha1.PhaseFailed:
+		klog.Infof("Reservation %s failed", reservation.Name)
+		return ctrl.Result{}, nil
+	}
 
```

Comment:
All this logic not strictly correspond with the one discussed yesterday, we need to talk again. It is surely better than before but we need to work a little bit more on that

Reply:
New logic implemented